### ISSUE TITLE
Critical error: login loop after update to 18.09.3 on ubuntu 18.10

### ISSUE THREAD

**Issue by Hunternnm** [docker/for-linux#612]
* [x] This is a bug report
* [ ] This is a feature request
* [x] I searched existing issues before opening this one

After upgrading to version 18.09.3 could not login. Since could not find the problem reinstalled the operating system. Everything worked as it should, but then I installed the docker and the problem reappeared. Through the second terminal, I deleted the docker and was able to log into my account. I checked it 5 times, and every time when I installed the docker, the login loop was repeated, when I deleted it - the login loop was returned. Changed only installed docker or not. I could not find any logs, can you tell me where to look. Installed on my laptop - the same result. A fresh installed operating system and docker result in a login loop.

PS The problem only occurs if you execute `sudo usermod -aG docker my-user`

### Actual behavior
Login loop after update to v18.09.3 on ubuntu 18.10

### Steps to reproduce the behavior

**Output of `docker version`:**

```
Client:
 Version:           18.09.3
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        774a1f4
 Built:             Thu Feb 28 06:47:20 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       774a1f4
  Built:            Thu Feb 28 05:59:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

**Output of `docker info`:**

```
Containers: 7
 Running: 0
 Paused: 0
 Stopped: 7
Images: 19
Server Version: 18.09.3
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: e6b3f5632f50dbc4e9cb6288d911bf4f5e95b18e
runc version: 6635b4f0c6af3810594d2770f662f34ddc15b40d
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.18.0-15-generic
Operating System: Ubuntu 18.10
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.789GiB
Name: home
ID: NJ5O:PPRQ:VCPP:YHR6:BF22:44IC:RQUE:LZJ2:YP2N:K4KF:4ZGN:NDVZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support

```

**Additional environment details (AWS, VirtualBox, physical, etc.)**

physical, AMD phenom  955, 8gb ram

**Comment by Hunternnm** [comment]
By the way, after login, when you try to turn off the message that there are active user sessions, although the login form is still on the screen. Maybe this additional information will help.

**Comment by seemethere** [comment]
Couldn't reproduce this with a local Vagrant VM:

<details>
<summary> OS Information </summary>

```
vagrant@ubuntu-cosmic:~$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.10 (Cosmic Cuttlefish)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.10"
VERSION_ID="18.10"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=cosmic
UBUNTU_CODENAME=cosmic
vagrant@ubuntu-cosmic:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.10
Release:	18.10
Codename:	cosmic
vagrant@ubuntu-cosmic:~$ docker version
Client:
 Version:           18.09.3
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        774a1f4
 Built:             Thu Feb 28 06:47:20 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.3
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       774a1f4
  Built:            Thu Feb 28 05:59:55 2019
  OS/Arch:          linux/amd64
  Experimental:     false
vagrant@ubuntu-cosmic:~$ docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.09.3
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: e6b3f5632f50dbc4e9cb6288d911bf4f5e95b18e
runc version: 6635b4f0c6af3810594d2770f662f34ddc15b40d
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.18.0-15-generic
Operating System: Ubuntu 18.10
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 985.1MiB
Name: ubuntu-cosmic
ID: UVTO:E5TW:RAMC:SCIN:VVYH:5HFG:EQTW:2A2U:IS3L:VQYJ:RKXH:JVZP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: seemethere
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine

WARNING: No swap limit support
```

</details>

**Comment by Hunternnm** [comment]
@seemethere Hmm, I can shoot a video where I repeat all my steps. Logs are empty(

**Comment by seemethere** [comment]
@Hunternnm that'd be ideal, thanks!

**Comment by Hunternnm** [comment]
@seemethere sorry for this very stupid video and my trembling hands https://photos.app.goo.gl/L2dvsmz14Dq1u5Ma9

**Comment by Hunternnm** [comment]
I found the problem, but it is rather strange. The problem was that the container could not start with an error. Under the link it can be found on a code section `Failed to start container cbef44b030fc7ae4faceb9f1bc70431660155f27d87caafd1cd85155996bba42`.

there is this line in my docker-compose file  that has broken everything
```
volumes:
  - $SSH_AUTH_SOCK:/ssh-agent
```

https://gist.github.com/Hunternnm/7adf1eea973b9a18537e3478cd7f96c3

**Comment by thaJeztah** [comment]
The error is this:


```
mounting "/run/user/1000/keyring/ssh" to rootfs "/var/lib/docker/overlay2/69a6b9254893212e2a6c75c27dba8b86a1c8e2ad7ec52ac1bc46ce0059acb9fc/merged"
at "/var/lib/docker/overlay2/69a6b9254893212e2a6c75c27dba8b86a1c8e2ad7ec52ac1bc46ce0059acb9fc/merged/ssh-agent" caused "not a directory":
unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
```

This is likely caused by a race-condition, where `ssh-agent` is not yet running, thus `$SSH_AUTH_SOCK` is not yet present, but docker is already starting the container.


That's a known problem when using the shorthand `-v <hostpath>:<containerpath>` syntax to bind mount a file from the host; when using that syntax, Docker will look for `<hostpath`, and in case the path is not found, will auto-create the host-path.  

Given that (with just the "path" known), Docker does not know if you're attempting to bind-mount a _file_, a _socket_, or a _directory_, it _assumes_ you want to bind-mount a _directory_ , and create a directory at the given path. Bind-mounting a directory over a file (inside the container) is not possible, so the container will fail to start.

However, in your situation, the problem is likely causing more problems, because now  that docker has created a directory at `$SSH_AUTH_SOCK`, `ssh-agent` may no longer be able to start (it cannot create the socket, because a directory exists in that location).


We tried deprecating the "auto-create" functionality when using the `-v` flag, but unfortunately, there's too many users relying on that feature, so removing that functionality would be a breaking change for many (https://github.com/moby/moby/pull/19953#issuecomment-198024624), and we had to revert that change (https://github.com/moby/moby/pull/21666).

Instead, a new `--mount` flag was added that addresses this (and quite some other) issues. The `--mount` flag is more "strict", and does _not_ automatically create the host-path, and instead will produce an error when trying to start the container.


So, to prevent this problem from occuring; (first be sure to remove the directory that was created)

Then, instead of using the shorthand `-v` syntax, use the `--mount` syntax; 

```bash
--mount type=bind,source=$SSH_AUTH_SOCK,target=/ssh-agent
```


Or, in your docker compose file, use the [long form syntax](https://docs.docker.com/compose/compose-file/#long-syntax-3) for the volume:

```yaml
- type: bind
  source: $SSH_AUTH_SOCK
  target: /ssh-agent
```
